### ISSUE TITLE
The mismatched types were not detected, and a segmentation fault occurred at runtime.

### ISSUE THREAD

**Issue by rucsesec** [rust-lang/rust#139792]
I tried this code:

```rust
use std::fmt::Display;

trait Trait{
    type Associated;
}

impl<T: ?Sized> Trait for T {
    type Associated = Box<dyn Display>;
}

fn foo<T: ?Sized>(_: <T as Trait>::Associated) {}

fn main() {
    let x = "Hello World".to_string();
    foo::<dyn Trait<Associated = &str>>( &x );
}
```
I expected to see this happen: *mismatched types：expected “Box\<dyn Display\>”， found “&str”*.
Instead, this happened: *The code compiles successfully but encounters a segmentation fault at runtime.*

```shell
Program returned: 139
Program stderr
Program terminated with signal: SIGSEGV
```

Demo: [godbolt permalink](https://godbolt.org/z/KhzfYbzqc)
Demo: [on current playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=af5a55ba752a8433f9d8b405670e5cd9)
### Meta

`rustc --version --verbose`:
```
rustc 1.88.0-nightly (0fe8f3454 2025-04-10)
```

**Comment by jieyouxu** [comment]
This also segfaults on stable
